File: azure-communication-ui/calling/src/test/java/com/azure/android/communication/ui/calling/redux/middleware/AudioDeviceMiddlewareTest.java
package com.azure.android.communication.ui.calling.redux.middleware;

import android.os.Handler;
import android.os.Looper;

import com.azure.android.communication.ui.calling.redux.Store;
import com.azure.android.communication.ui.calling.redux.action.AudioDeviceAction;
import com.azure.android.communication.ui.calling.redux.state.AudioDevice;
import com.azure.android.communication.ui.calling.redux.state.AudioDeviceState;
import com.azure.android.communication.ui.calling.redux.state.AudioDeviceType;
import com.azure.android.communication.ui.calling.service.AudioDeviceService;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.annotation.LooperMode;

import java.util.Arrays;

import static org.mockito.Mockito.*;
import static org.junit.Assert.*;

@RunWith(RobolectricTestRunner.class)
@LooperMode(LooperMode.Mode.PAUSED)
public class AudioDeviceMiddlewareTest {
    @Mock
    private Store mockStore;
    @Mock
    private AudioDeviceService mockAudioDeviceService;
    @Mock
    private AudioDeviceMiddleware.Dispatch mockNext;

    private AudioDeviceMiddleware middleware;
    private AudioDeviceState initialState;
    private AudioDevice speaker;
    private AudioDevice wiredHeadset;
    private AudioDevice bluetooth;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        middleware = new AudioDeviceMiddleware(mockAudioDeviceService);

        speaker = new AudioDevice(AudioDeviceType.SPEAKER, "speaker", "Speaker");
        wiredHeadset = new AudioDevice(AudioDeviceType.WIRED_HEADSET, "wired_headset", "Wired Headset");
        bluetooth = new AudioDevice(AudioDeviceType.BLUETOOTH, "bluetooth", "Bluetooth");

        initialState = new AudioDeviceState(
            speaker,
            Arrays.asList(speaker, wiredHeadset, bluetooth),
            true,
            true,
            com.azure.android.communication.ui.calling.redux.state.AudioDeviceSwitchingState.NONE,
            null
        );

        when(mockStore.getCurrentState()).thenReturn(initialState);
    }

    @Test
    public void testSelectDeviceRequested_Success() {
        when(mockAudioDeviceService.configureAudioRouting(wiredHeadset)).thenReturn(true);

        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(wiredHeadset), mockNext);

        // Verify DeviceSwitchStarted action is dispatched immediately
        ArgumentCaptor<Object> actionCaptor = ArgumentCaptor.forClass(Object.class);
        verify(mockNext, times(1)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getValue() instanceof AudioDeviceAction.DeviceSwitchStarted);

        // Execute pending Looper tasks
        Shadows.shadowOf(Looper.getMainLooper()).idle();

        // Verify DeviceSwitchCompleted action is dispatched after async operation
        verify(mockNext, times(2)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getAllValues().get(1) instanceof AudioDeviceAction.DeviceSwitchCompleted);
    }

    @Test
    public void testSelectDeviceRequested_Failure() {
        when(mockAudioDeviceService.configureAudioRouting(bluetooth)).thenReturn(false);

        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(bluetooth), mockNext);

        // Verify DeviceSwitchStarted action is dispatched immediately
        ArgumentCaptor<Object> actionCaptor = ArgumentCaptor.forClass(Object.class);
        verify(mockNext, times(1)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getValue() instanceof AudioDeviceAction.DeviceSwitchStarted);

        // Execute pending Looper tasks
        Shadows.shadowOf(Looper.getMainLooper()).idle();

        // Verify DeviceSwitchFailed action is dispatched after async operation
        verify(mockNext, times(2)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getAllValues().get(1) instanceof AudioDeviceAction.DeviceSwitchFailed);

        // Verify fallback to previous device is attempted
        verify(mockAudioDeviceService).configureAudioRouting(speaker);
    }

    @Test
    public void testSelectDeviceRequested_UnavailableDevice() {
        AudioDevice unavailableDevice = new AudioDevice(AudioDeviceType.BLUETOOTH, "unavailable", "Unavailable");

        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(unavailableDevice), mockNext);

        // Verify DeviceSwitchFailed action is dispatched immediately
        ArgumentCaptor<Object> actionCaptor = ArgumentCaptor.forClass(Object.class);
        verify(mockNext, times(1)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getValue() instanceof AudioDeviceAction.DeviceSwitchFailed);

        // Verify no device configuration is attempted
        verify(mockAudioDeviceService, never()).configureAudioRouting(any());
    }

    @Test
    public void testSelectDeviceRequested_AlreadySelected() {
        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(speaker), mockNext);

        // Verify no actions are dispatched
        verify(mockNext, never()).dispatch(any());
        verify(mockAudioDeviceService, never()).configureAudioRouting(any());
    }

    @Test
    public void testSelectDeviceRequested_AlreadySwitching() {
        AudioDeviceState switchingState = new AudioDeviceState(
            speaker,
            Arrays.asList(speaker, wiredHeadset, bluetooth),
            true,
            true,
            com.azure.android.communication.ui.calling.redux.state.AudioDeviceSwitchingState.SWITCHING,
            null
        );
        when(mockStore.getCurrentState()).thenReturn(switchingState);

        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(bluetooth), mockNext);

        // Verify no actions are dispatched
        verify(mockNext, never()).dispatch(any());
        verify(mockAudioDeviceService, never()).configureAudioRouting(any());
    }

    @Test
    public void testSelectDeviceRequested_Exception() {
        when(mockAudioDeviceService.configureAudioRouting(bluetooth))
            .thenThrow(new RuntimeException("Test error"));

        middleware.process(mockStore, new AudioDeviceAction.SelectDeviceRequested(bluetooth), mockNext);

        // Verify DeviceSwitchStarted action is dispatched immediately
        ArgumentCaptor<Object> actionCaptor = ArgumentCaptor.forClass(Object.class);
        verify(mockNext, times(1)).dispatch(actionCaptor.capture());
        assertTrue(actionCaptor.getValue() instanceof AudioDeviceAction.DeviceSwitchStarted);

        // Execute pending Looper tasks
        Shadows.shadowOf(Looper.getMainLooper()).idle();

        // Verify DeviceSwitchFailed action is dispatched after async operation
        verify(mockNext, times(2)).dispatch(actionCaptor.capture());
        Object failedAction = actionCaptor.getAllValues().get(1);
        assertTrue(failedAction instanceof AudioDeviceAction.DeviceSwitchFailed);
        assertEquals("Error configuring audio routing: Test error",
            ((AudioDeviceAction.DeviceSwitchFailed) failedAction).getError());

        // Verify fallback to previous device is attempted
        verify(mockAudioDeviceService).configureAudioRouting(speaker);
    }
}

File: azure-communication-ui/calling/src/test/java/com/azure/android/communication/ui/calling/redux/middleware/AudioDeviceMiddlewareTest.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.android.communication.ui.calling.redux.middleware

import android.os.Looper
import com.azure.android.communication.ui.calling.logger.Logger
import com.azure.android.communication.ui.calling.redux.Store
import com.azure.android.communication.ui.calling.redux.action.Action
import com.azure.android.communication.ui.calling.redux.action.AudioDeviceAction
import com.azure.android.communication.ui.calling.redux.state.AudioDevice
import com.azure.android.communication.ui.calling.redux.state.AudioDeviceState
import com.azure.android.communication.ui.calling.redux.state.AudioDeviceType
import com.azure.android.communication.ui.calling.redux.state.ReduxState
import com.azure.android.communication.ui.calling.service.AudioDeviceService

import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows.shadowOf
import org.robolectric.annotation.LooperMode

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.PAUSED)
class AudioDeviceMiddlewareTest {
    @Mock
    private lateinit var mockStore: Store<ReduxState>
    @Mock
    private lateinit var mockAudioDeviceService: AudioDeviceService
    @Mock
    private lateinit var mockLogger: Logger
    @Mock
    private lateinit var mockNext: (Action) -> Any

    private lateinit var middleware: AudioDeviceMiddlewareImpl
    private lateinit var state: ReduxState
    private val speaker = AudioDevice(AudioDeviceType.SPEAKER, "speaker", "Speaker")
    private val wiredHeadset = AudioDevice(AudioDeviceType.WIRED_HEADSET, "wired_headset", "Wired Headset")
    private val bluetooth = AudioDevice(AudioDeviceType.BLUETOOTH, "bluetooth", "Bluetooth")

    @Test
    fun `test select device requested success`() {
        setupTest()
        `when`(mockAudioDeviceService.configureAudioRouting(wiredHeadset)).thenReturn(true)

        val action = AudioDeviceAction.SelectDeviceRequested(wiredHeadset)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify DeviceSwitchStarted action is dispatched immediately
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchStarted::class.java))

        // Execute pending Looper tasks
        shadowOf(Looper.getMainLooper()).idle()

        // Verify DeviceSwitchCompleted action is dispatched after async operation
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchCompleted::class.java))
    }

    @Test
    fun `test select device requested failure`() {
        setupTest()
        `when`(mockAudioDeviceService.configureAudioRouting(bluetooth)).thenReturn(false)

        val action = AudioDeviceAction.SelectDeviceRequested(bluetooth)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify DeviceSwitchStarted action is dispatched immediately
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchStarted::class.java))

        // Execute pending Looper tasks
        shadowOf(Looper.getMainLooper()).idle()

        // Verify DeviceSwitchFailed action is dispatched after async operation
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchFailed::class.java))

        // Verify fallback to previous device is attempted
        verify(mockAudioDeviceService).configureAudioRouting(speaker)
    }

    @Test
    fun `test select device requested unavailable device`() {
        setupTest()
        val unavailableDevice = AudioDevice(AudioDeviceType.BLUETOOTH, "unavailable", "Unavailable")

        val action = AudioDeviceAction.SelectDeviceRequested(unavailableDevice)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify DeviceSwitchFailed action is dispatched immediately
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchFailed::class.java))

        // Verify no device configuration is attempted
        verify(mockAudioDeviceService, never()).configureAudioRouting(any())
    }

    @Test
    fun `test select device requested already selected`() {
        setupTest()

        val action = AudioDeviceAction.SelectDeviceRequested(speaker)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify no actions are dispatched
        verify(mockNext, never()).invoke(any())
        verify(mockAudioDeviceService, never()).configureAudioRouting(any())
    }

    @Test
    fun `test select device requested already switching`() {
        setupTest(switching = true)

        val action = AudioDeviceAction.SelectDeviceRequested(bluetooth)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify no actions are dispatched
        verify(mockNext, never()).invoke(any())
        verify(mockAudioDeviceService, never()).configureAudioRouting(any())
    }

    @Test
    fun `test select device requested exception`() {
        setupTest()
        `when`(mockAudioDeviceService.configureAudioRouting(bluetooth))
            .thenThrow(RuntimeException("Test error"))

        val action = AudioDeviceAction.SelectDeviceRequested(bluetooth)
        middleware.invoke(mockStore)(mockNext)(action)

        // Verify DeviceSwitchStarted action is dispatched immediately
        verify(mockNext).invoke(any(AudioDeviceAction.DeviceSwitchStarted::class.java))

        // Execute pending Looper tasks
        shadowOf(Looper.getMainLooper()).idle()

        // Verify DeviceSwitchFailed action is dispatched after async operation
        verify(mockNext).invoke(
            argThat { arg ->
                arg is AudioDeviceAction.DeviceSwitchFailed &&
                arg.error == "Error configuring audio routing: Test error"
            }
        )

        // Verify fallback to previous device is attempted
        verify(mockAudioDeviceService).configureAudioRouting(speaker)
    }

    private fun setupTest(switching: Boolean = false) {
        mockStore = mock(Store::class.java) as Store<ReduxState>
        mockAudioDeviceService = mock(AudioDeviceService::class.java)
        mockLogger = mock(Logger::class.java)
        mockNext = mock((Action) -> Any::class.java)

        middleware = AudioDeviceMiddlewareImpl(mockAudioDeviceService, mockLogger)

        val audioState = AudioDeviceState(
            speaker,
            listOf(speaker, wiredHeadset, bluetooth),
            true,
            true,
            if (switching) com.azure.android.communication.ui.calling.redux.state.AudioDeviceSwitchingState.SWITCHING
            else com.azure.android.communication.ui.calling.redux.state.AudioDeviceSwitchingState.NONE,
            null
        )

        state = mock(ReduxState::class.java)
        `when`(state.audioState).thenReturn(audioState)
        `when`(mockStore.getCurrentState()).thenReturn(state)
    }

    private fun <T> any(type: Class<T>): T = any()
}
